File: server/src/main/java/com/backendProject/SoulSync/user/model/UserModel.java
package com.backendProject.SoulSync.user.model;

import com.backendProject.SoulSync.enums.UserStatus;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
public class UserModel {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;
    @Column(nullable = false, unique = true)
    private String username;
    @Column(nullable = false, unique = true)
    private String email;

    @Column(nullable = false)
    private String password;

    // Profile fields
    private Integer age;
    private String name;
    private String gender;
    private String bio;
    private String location;
    private String interests;
    private String profileImageUrl;
    //    other important fields for dating
    private Double height;
    private String sports;
    private String games;
    private String relationshipType; //casual or commitment
    private Boolean goesGym;
    private Boolean shortHair;
    private Boolean wearGlasses;
    private Boolean drink;
    private Boolean smoke;
    //deactivation fields
    private Boolean isActive = true;
    // Reason codes 1 = User requested delete, 2 = Banned after reports, 3 = Misc
    @Column(nullable = true)
    private Integer deactivationReason; // null means active

    private LocalDateTime deactivatedAt;

    @Column(nullable = false)
    private Integer reportCount = 0;

    @CreationTimestamp
    private LocalDateTime createdAt;


    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private UserStatus status = UserStatus.OFFLINE;

    @ManyToMany
    @JoinTable(
            name = "user_likes",
            joinColumns = @JoinColumn(name = "liker_id"),
            inverseJoinColumns = @JoinColumn(name = "liked_id")
    )
    @JsonIgnore
    private Set<UserModel> likedUsers = new HashSet<>();

    public UserStatus getStatus() {
        return status;
    }

    public void setStatus(UserStatus status) {
        this.status = status;
    }

    public Set<UserModel> getLikedUsers() {
        return likedUsers;
    }

    public void setLikedUsers(Set<UserModel> likedUsers) {
        this.likedUsers = likedUsers;
    }

    public Double getHeight() {
        return height;
    }

    public void setHeight(Double height) {
        this.height = height;
    }

    public String getSports() {
        return sports;
    }

    public void setSports(String sports) {
        this.sports = sports;
    }

    public String getGames() {
        return games;
    }

    public void setGames(String games) {
        this.games = games;
    }

    public String getRelationshipType() {
        return relationshipType;
    }

    public void setRelationshipType(String relationshipType) {
        this.relationshipType = relationshipType;
    }

    public Boolean getGoesGym() {
        return goesGym;
    }

    public void setGoesGym(Boolean goesGym) {
        this.goesGym = goesGym;
    }

    public Boolean getShortHair() {
        return shortHair;
    }

    public void setShortHair(Boolean shortHair) {
        this.shortHair = shortHair;
    }

    public Boolean getWearGlasses() {
        return wearGlasses;
    }

    public void setWearGlasses(Boolean wearGlasses) {
        this.wearGlasses = wearGlasses;
    }

    public Boolean getDrink() {
        return drink;
    }

    public void setDrink(Boolean drink) {
        this.drink = drink;
    }

    public Boolean getSmoke() {
        return smoke;
    }

    public void setSmoke(Boolean smoke) {
        this.smoke = smoke;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Boolean getActive() {
        return isActive;
    }

    public void setActive(Boolean active) {
        isActive = active;
    }

    public Integer getAge() {
        return age;
    }

    public void setAge(Integer age) {
        this.age = age;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getBio() {
        return bio;
    }

    public void setBio(String bio) {
        this.bio = bio;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public String getInterests() {
        return interests;
    }

    public void setInterests(String interests) {
        this.interests = interests;
    }

    public String getProfileImageUrl() {
        return profileImageUrl;
    }

    public void setProfileImageUrl(String profileImageUrl) {
        this.profileImageUrl = profileImageUrl;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }


    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public Integer getDeactivationReason() {
        return deactivationReason;
    }

    public void setDeactivationReason(Integer deactivationReason) {
        this.deactivationReason = deactivationReason;
    }

    public LocalDateTime getDeactivatedAt() {
        return deactivatedAt;
    }

    public void setDeactivatedAt(LocalDateTime deactivatedAt) {
        this.deactivatedAt = deactivatedAt;
    }

    public Integer getReportCount() {
        return reportCount;
    }

    public void setReportCount(Integer reportCount) {
        this.reportCount = reportCount;
    }
}

File: lib/app/data/models/user_model.dart
class UserModel {
  final int? id;
  final String? name;
  final int? age;
  final String? gender;
  final String? bio;
  final String? location;
  final String? interests;
  final String? profileImageUrl;
  final bool? isLikedByMe;
  final double? height;
  final String? sports;
  final String? games;
  final String? relationshipType;
  final bool? goesGym;
  final bool? shortHair;
  final bool? wearGlasses;
  final bool? drink;
  final bool? smoke;
  final int? score;

  UserModel({
    this.id,
    this.name,
    this.age,
    this.gender,
    this.bio,
    this.location,
    this.interests,
    this.profileImageUrl,
    this.isLikedByMe,
    this.height,
    this.sports,
    this.games,
    this.relationshipType,
    this.goesGym,
    this.shortHair,
    this.wearGlasses,
    this.drink,
    this.smoke,
    this.score,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'],
      name: json['name'],
      age: json['age'],
      gender: json['gender'],
      bio: json['bio'],
      location: json['location'],
      interests: json['interests'],
      profileImageUrl: json['profileImageUrl'],
      isLikedByMe: json['isLikedByMe'] ?? false,
      height: json['height']?.toDouble(),
      sports: json['sports'],
      games: json['games'],
      relationshipType: json['relationshipType'],
      goesGym: json['goesGym'],
      shortHair: json['shortHair'],
      wearGlasses: json['wearGlasses'],
      drink: json['drink'],
      smoke: json['smoke'],
      score: json['score'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'age': age,
      'gender': gender,
      'bio': bio,
      'location': location,
      'interests': interests,
      'profileImageUrl': profileImageUrl,
      'height': height,
      'sports': sports,
      'games': games,
      'relationshipType': relationshipType,
      'goesGym': goesGym,
      'shortHair': shortHair,
      'wearGlasses': wearGlasses,
      'drink': drink,
      'smoke': smoke,
    };
  }

  UserModel copyWith({
    int? id,
    String? name,
    int? age,
    String? gender,
    String? bio,
    String? location,
    String? interests,
    String? profileImageUrl,
    bool? isLikedByMe,
    double? height,
    String? sports,
    String? games,
    String? relationshipType,
    bool? goesGym,
    bool? shortHair,
    bool? wearGlasses,
    bool? drink,
    bool? smoke,
    int? score,
  }) {
    return UserModel(
      id: id ?? this.id,
      name: name ?? this.name,
      age: age ?? this.age,
      gender: gender ?? this.gender,
      bio: bio ?? this.bio,
      location: location ?? this.location,
      interests: interests ?? this.interests,
      profileImageUrl: profileImageUrl ?? this.profileImageUrl,
      isLikedByMe: isLikedByMe ?? this.isLikedByMe,
      height: height ?? this.height,
      sports: sports ?? this.sports,
      games: games ?? this.games,
      relationshipType: relationshipType ?? this.relationshipType,
      goesGym: goesGym ?? this.goesGym,
      shortHair: shortHair ?? this.shortHair,
      wearGlasses: wearGlasses ?? this.wearGlasses,
      drink: drink ?? this.drink,
      smoke: smoke ?? this.smoke,
      score: score ?? this.score,
    );
  }
}